File: minesweeper/java-minesweeper/src/main/java/it/chicio/minesweeper/field/formatter/FieldsFormatterWithHeader.java
package it.chicio.minesweeper.field.formatter;

import it.chicio.minesweeper.field.Field;

import java.util.List;

public class FieldsFormatterWithHeader implements FieldsFormatter {
    private FieldFormatter fieldFormatter;

    public FieldsFormatterWithHeader(FieldFormatter fieldFormatter) {
        this.fieldFormatter = fieldFormatter;
    }

    public String format(List<Field> fields) {
        StringBuilder formattedField = new StringBuilder();
        int currentFieldPosition = 1;
        for (Field field : fields) {
            formattedField.append(createHeaderFor(currentFieldPosition)).append(fieldFormatter.format(field));
            appendNewlineIfNeeded(fields, formattedField, currentFieldPosition);
            currentFieldPosition++;
        }
        return formattedField.toString();
    }

    private void appendNewlineIfNeeded(List<Field> fields, StringBuilder formattedField, int currentFieldPosition) {
        if (currentFieldPosition != fields.size()) {
            formattedField.append(System.getProperty("line.separator"));
        }
    }

    private String createHeaderFor(int currentField) {
        return "field #" + currentField + ":" + System.getProperty("line.separator");
    }
}

File: minesweeper/kotlin-mockk-minesweeper/src/main/kotlin/it/chicio/minesweeper/field/formatter/FieldsFormatterWithHeader.kt
package it.chicio.minesweeper.field.formatter

import it.chicio.minesweeper.field.Field

class FieldsFormatterWithHeader(private val fieldFormatter: FieldFormatter) : FieldsFormatter {
    override fun format(fields: List<Field>): String = fields
            .mapIndexed { index, field -> "field #${index + 1}:\n".plus(fieldFormatter.format(field)) }
            .joinToString(System.getProperty("line.separator"))
}